### PR TITLE
Make adding callbacks thread-safe

### LINKED ISSUE #7219: Tornado Next Tick Callback Issue
Posted this to the Google Group a couple weeks ago but didn't seem to get any activity there but wanted to bring this to attention:

Build is on bokeh 0.12.9, tornado version 4.5.2. Chrome browser. 

I've been building an application to consume a UDP stream and add the data as it comes in to a bokeh plot. I can consume the UDP stream fine and most of the time the bokeh callbacks work to add the data but sometimes I get a long error message as follows:

2017-10-25 14:35:52,825 Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x000000ADC0CA3048>)
Traceback (most recent call last):
  File "d:\afcbm\continuum\lib\site-packages\tornado\ioloop.py", line 605, in _run_callback
    ret = callback()
  File "d:\afcbm\continuum\lib\site-packages\tornado\stack_context.py", line 277, in null_wrapper
    return fn(*args, **kwargs)
  File "d:\afcbm\continuum\lib\site-packages\bokeh\util\tornado.py", line 133, in wrapper
    self.remove_next_tick_callback(callback)
  File "d:\afcbm\continuum\lib\site-packages\bokeh\util\tornado.py", line 160, in remove_next_tick_callback
    self._remove(callback, self._next_tick_callbacks)
  File "d:\afcbm\continuum\lib\site-packages\bokeh\util\tornado.py", line 155, in _remove
    self._error_on_double_remove(callback, callbacks)
  File "d:\afcbm\continuum\lib\site-packages\bokeh\util\tornado.py", line 116, in _error_on_double_remove
    raise ValueError("Removing a callback twice (or after it's already been run)")
ValueError: Removing a callback twice (or after it's already been run)

Minimal Code example below:

```
import socket
import struct
import time
from bokeh.plotting import figure
from bokeh.layouts import column
from bokeh.models import ColumnDataSource
from bokeh.io import curdoc
from threading import Thread
from functools import partial
from tornado import gen

plot_column = column()

UDP_IP = "127.0.0.1" # 172.17.1.32
UDP_PORT = 29505

sock = socket.socket(socket.AF_INET, # Internet
                     socket.SOCK_DGRAM) # UDP

sock.bind((UDP_IP, UDP_PORT))

doc = curdoc()

f = figure(plot_width=700, plot_height=200)
source = ColumnDataSource(data=dict(x=[], y=[]))
l = f.circle(x='x', y='y', source=source)
plot_column.children.insert(0, f)

@gen.coroutine
def update(x, val):
    source.stream(dict(x=[x], y=[val]))


def receive_loop():
    while True:
        data, addr = sock.recvfrom(1024) # buffer size
        val = struct.unpack('!i', data)[0]
        doc.add_next_tick_callback(partial(update, x=time.time(), val=val))


doc.add_root(plot_column)
thread = Thread(target=receive_loop)
thread.start()


Minimal code example for the client:

import socket
from random import random
import time
import struct

UDP_IP = "127.0.0.1" # 172.17.1.32
UDP_PORT = 29505

sock = socket.socket(socket.AF_INET,
                     socket.SOCK_DGRAM)
x = 0
while True:
    time.sleep(0.2)
    x += 1

    sock.sendto(struct.pack('!i', x), (UDP_IP, UDP_PORT))

```

Interestingly, I've been able to create a temporary workaround on my system by modifying the tornado.py file within bokeh by adding a time.sleep(.001) call at line 154 within the _remove method. With any amount of sleep in the process of removing the next tick callback the ValueError shown above doesn't trigger. The error seems to show randomly on the system, with no real consistency of when it shows up within the stream of data.

### PR DESCRIPTION
**Pull request by p-himik**
Fix #7219

### CODE REVIEW

**Comment by bryevdv** [comment]
This will need a migration note since it is a breaking change.

**Comment by p-himik** [comment]
@bryevdv Where should I add it?

**Comment by bryevdv** [comment]
@p-himik release notes are here:

https://github.com/bokeh/bokeh/blob/master/sphinx/source/docs/releases/0.12.15.rst

**Comment by bryevdv** [comment]
@p-himik just trying to see if there is any way to not have an immediately breaking change. What if the "add" functions maintain a map of functions to id's returned for that function (i.e. if you add `foo` twice, then `foo` maps to two ids). And then we still allow removing by value, and when that is done the remove code is called for all the ids? We could issue a deprecation warning in this case too.

**Comment by p-himik** [comment]
@bryevdv I think it would complicate code that's already complicated enough, and for no significant gain.
I sincerely doubt that even 1% of Bokeh users use `remove_*_callback` functions. And for those who do use them, the migration process w.r.t. this change is hardly more involved than changing `bokeh==0.12.14` to `bokeh==0.12.15`.

**Comment by bryevdv** [comment]
I agree there are probably not that many users, but 1% of users would still be several thousand people, and I'm the one who gets dragged for breaking changes. A good part of the gain in this instance would be my peace of mind, which is significant to me, at least :)

**Comment by p-himik** [comment]
Hmm ok, sure - I'll add it. But I must admit to me it looks kind of strange given that there's no deprecation of programmatically settable selection.

**Comment by bryevdv** [comment]
>  that there's no deprecation of programmatically settable selection.

It's nontechnically possible on the JS side, AFAIK. And I'm not sure how it could be done on the python side offhand either. The original design with a "raw" dict was bad and limiting in many way, including the ability to move away from it if needed. Obviously it should've been more abstracted from the start to insulate from changes, and failing to do that is my fault.

**Comment by p-himik** [comment]
Ah, I see. Thanks for explaining.

**Comment by bryevdv** [comment]
@p-himik this was a it more work than I thought it would be, thank you for taking the time, I appreciate it. It's good for us to start getting into the habit of being diligent about changes, as that will be much more paramount after 1.0

File: bokeh/util/tornado.py [hunk 1]
```diff
@@ -100,103 +112,133 @@ def __init__(self, io_loop=None):
...
+            if callback_id is None:
+                callback_id = make_id()
+            elif callback_id in removers:
+                raise ValueError("A callback of the same type has already been added with this ID")
+            removers[callback_id] = remover
+            self._assign_deprecated_remover(callback, callback_id, removers)
+            return callback_id
+
+    def _create_deprecated_remover(self, callback, removers):
+        deprecation.deprecated((0, 12, 14),
```
Comment (bryevdv) [inline bokeh/util/tornado.py]: should be `(0, 12, 15)`

File: bokeh/document/document.py [hunk 2]
```diff
@@ -967,9 +979,19 @@ def _recompute_all_models(self):
         self._all_models = recomputed
         self._all_models_by_name = recomputed_by_name
 
-    def _remove_session_callback(self, callback):
-        ''' Remove a callback added earlier with ``add_periodic_callback``
-        or ``add_timeout_callback``.
+    def _deprecated_remove_session_callback(self, callback, originator):
+        deprecation.deprecated((0, 12, 14),
```
Comment (bryevdv) [inline bokeh/document/document.py]: same, should be `(0, 12, 15)`

**Comment by bryevdv** [comment]
Thanks @p-himik , will merge when green!

**Comment by bryevdv** [comment]
Hrm, you may need to rebase on `master` after the hexbin PR merge @p-himik , sorry for the hassle

**Comment by bryevdv** [comment]
I want to get this merged and then cut another dev build so I went ahead and rebased!

**Pull request merged**